uri.go

```diff
@@ -0,0 +1,201 @@
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+		return parser.Parse(ctx, f)
+	case "git", "git+https", "git+ssh":
+		return parseModuleGitURI(ctx, cln, u.Scheme, uri)
+	default:
+		return nil, fmt.Errorf("%q is not a valid module uri scheme", u.Scheme)
+	}
+}
+
+func parseModuleGitURI(ctx context.Context, cln *client.Client, scheme, uri string) (*ast.Module, error) {
+	u, err := gitscheme.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if u.Filename == "" {
+		u.Filename = DefaultFilename
+	}
+
+	user := u.User
+	if user == "" {
+		user = "git"
+	}
+
+	ssh := false
+	sockPath := local.Env(ctx, "SSH_AUTH_SOCK")
+	switch scheme {
+	case "git+https": // explicit https
+	case "git+ssh": // explicit ssh
+		err = testSSHAgent(sockPath, u.Host, user)
+		if err != nil {
+			return nil, err
+		}
+		ssh = true
+	case "git": // auto
+		if sockPath != "" {
+			err = testSSHAgent(sockPath, u.Host, user)
+			if err == nil {
+				ssh = true
+			}
+			// TODO: HLB logging system.
+		}
+	}
+
+	var (
+		gitOpts     = []llb.GitOption{llb.KeepGitDir()}
+		sessionOpts []llbutil.SessionOption
+		remote      string
+		root        string
+	)
+	if !ssh {
+		root = u.Host + u.Path
+		remote = "https://" + root
+	} else {
+		// Use ssh protocol.
+		root = user + "@" + u.Host + u.Path
+		remote = "ssh://" + root
+
+		// Forward ssh agent.
+		sessionOpts = append(sessionOpts, llbutil.WithAgentConfig("default", sockproxy.AgentConfig{
+			ID:    "default",
+			SSH:   true,
+			Paths: []string{sockPath},
+		}))
+
+		keys, err := defaultKnownHostKeys()
+		if err != nil {
+			return nil, err
+		}
+		gitOpts = append(gitOpts, llb.KnownSSHHosts(keys))
+	}
+	if u.Branch != "" {
+		root = root + "@" + u.Branch
+	}
+
+	st := llb.Git(remote, u.Branch, gitOpts...)
+	def, err := st.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var pw progress.Writer
+	mw := codegen.MultiWriter(ctx)
+	if mw != nil {
+		pw = mw.WithPrefix(fmt.Sprintf("module %s", uri), true)
+	}
+
+	dir, err := solver.NewRemoteDirectory(ctx, cln, pw, def, root, "", nil, sessionOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	rc, err := dir.Open(u.Filename)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	mod, err := parser.Parse(ctx, rc)
+	if err != nil {
+		return nil, err
+	}
+	mod.Directory = dir
+	return mod, nil
+}
+
+func testSSHAgent(sockPath, host, user string) error {
+	conn, err := net.Dial("unix", sockPath)
+	if err != nil {
+		return err
+	}
+	sshAgent := agent.NewClient(conn)
+
+	knownHosts, err := defaultKnownHosts()
+	if err != nil {
+		return err
+	}
+
+	cfg := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeysCallback(sshAgent.Signers),
+		},
+		HostKeyCallback: knownHosts,
+	}
+	cln, err := ssh.Dial("tcp", host+":22", cfg)
+	if err != nil {
+		return err
+	}
+	defer cln.Close()
+
+	return nil
+}
+
+func defaultKnownHosts() (ssh.HostKeyCallback, error) {
+	filename, err := parser.ExpandHomeDir("~/.ssh/known_hosts")
```

Comment:
I think this will work correctly on a host that uses `\` separators. It might be more idiomatic to do something like `filepath.Join("~", ".ssh", "known_hosts")` though. Up to you.